File: src/foreign_if/python/examples/svr_demo.py
#!/usr/bin/env python

import sys
import numpy as np
np.set_printoptions(threshold=5)

#from sklearn.svm import LinearSVR
from frovedis.mllib.svm import LinearSVR

# initializing the Frovedis server
argvs = sys.argv
argc = len(argvs)
if (argc < 2):
    print ('Please give frovedis_server calling command as the first argument \n(e.g. "mpirun -np 2 /opt/nec/frovedis/ve/bin/frovedis_server")')
    quit()

from frovedis.exrpc.server import FrovedisServer
FrovedisServer.initialize(argvs[1])

# regression data
from sklearn.datasets import load_boston
mat, lbl = load_boston(return_X_y=True)

# fitting input matrix and label on svm regression object
svr = LinearSVR(loss='epsilon_insensitive', epsilon = 0.1,
                max_iter=1000, penalty='l2').fit(mat, lbl)

# predicting on loaded model
print("predicting on trained model: ")
print(svr.predict(mat))
print("prediction score: %.2f" % svr.score(mat, lbl))

# saving the model
svr.save("./out/SVRModel")

# loading the same model
svr.load("./out/SVRModel")

# debug_print
svr.debug_print()
svr.release()

FrovedisServer.shut_down()

File: src/foreign_if/spark/examples/scala/SVRDemo.scala
package test.scala;

import com.nec.frovedis.Jexrpc.FrovedisServer
import com.nec.frovedis.exrpc.FrovedisLabeledPoint
import com.nec.frovedis.mllib.regression.{SVRModel, LinearSVR, SVMRegressionWithSGD}
import com.nec.frovedis.matrix.FrovedisRowmajorMatrix
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.rdd.RDD
import org.apache.spark.mllib.linalg.{Vector, Vectors}
import org.apache.log4j.{Level, Logger}

object SVRDemo {
  def main(args: Array[String]): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)

    // -------- configurations --------
    val conf = new SparkConf().setAppName("SVRexample").setMaster("local[2]")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    // initializing Frovedis server with "personalized command", if provided in command line
    if(args.length != 0) FrovedisServer.initialize(args(0))

    // preparing sample train data
    val data = Vector(
       Vectors.dense(-1,1),
       Vectors.dense(-2, -1),
       Vectors.dense(-3, -2),
       Vectors.dense(1,1),
       Vectors.dense(2, 1),
       Vectors.dense(3, 2)
    )
    val lbl = Vector(10.0, 10.0, 10.0, 20.0, 10.0, 20.0)
    val zip1 = lbl zip data
    var lpv = zip1.map( a => LabeledPoint(a._1, a._2) ) // vector of LabeledPoint
    var d_lp: RDD[LabeledPoint] = sc.parallelize(lpv)  // distributed LabeledPoint

    // training hyper-parameters
    val numIter = 1000
    val stepSize = 0.01
    val loss = "epsilon_insensitive"
    val eps = 0.1
    val tol = 0.001
    val svr = new LinearSVR().setNumIter(numIter)
                             .setStepSize(stepSize)
                             .setLoss(loss)
                             .setEps(eps)
                             .setConvergenceTol(tol)
    val model = svr.run(d_lp)

    // to train with alreday constructed frovedis server side data
    //val f_lp = new FrovedisLabeledPoint(d_lp) // frovedis LabeledPoint
    //val model = svr.run(f_lp)

    var test_data = sc.parallelize(data)
    println("prediction with trained model: ")
    val pred = model.predict(test_data)
    pred.collect().foreach(println)

    // --- save/load ---
    model.save(sc, "out/mySVMRegressionModel")
    val sameModel = SVRModel.load("out/mySVMRegressionModel")

    println("prediction with loaded model: ")
    val pred2 = sameModel.predict(test_data)
    pred2.collect().foreach(println)

    // -------- clean-up --------
    model.release()
    sameModel.release()
    FrovedisServer.shut_down()
    sc.stop()
  }
}

